### PR TITLE
 ATIME-1037 - Improve performance of efforts table

### PR DESCRIPTION
**Pull request by andgeth**

### CODE REVIEW

File: arttime/arttime-webapp/src/main/java/com/artezio/arttime/filter/Filter.java [hunk 1]
```diff
@@ -174,11 +175,13 @@ public void setPredefined(Boolean predefined) {
...
 
+	//TODO rename
 	public List<Employee> getEmployeesInViewOfDepartments() {
-		return employees
-				.stream()
-				.filter(employee -> departments.contains(employee
-						.getDepartment())).collect(Collectors.toList());
+		return !containsAtLeastOneDepartment()
+				? employees
+				: employees.stream()
```
Comment (eshestakov) [inline arttime/arttime-webapp/src/main/java/com/artezio/arttime/filter/Filter.java]: possible NPE

File: arttime/arttime-webapp/src/main/java/com/artezio/arttime/web/spread_sheet/ProjectEffortsSpreadSheet.java [hunk 2]
```diff
@@ -111,7 +115,8 @@ protected boolean hasTeam(Project project, List<Project> subprojects) {
     @Override
     public boolean isHighLevel(SpreadSheetRow<?> row) {
 		Project project = row.getProject();
-		return project == null || !project.isSubproject() || !filter.getProjects().contains(project.getMaster());
+        return project == null || !project.isSubproject() ||
+                (!filter.getProjects().isEmpty() && !filter.getProjects().contains(project.getMaster()));
```
Comment (eshestakov) [inline arttime/arttime-webapp/src/main/java/com/artezio/arttime/web/spread_sheet/ProjectEffortsSpreadSheet.java]: possible NPE on filter.getProjects().* when projects == null

File: arttime/arttime-webapp/src/main/java/com/artezio/arttime/web/spread_sheet/SpreadSheet.java [hunk 3]
```diff
@@ -50,13 +59,27 @@ public SpreadSheet(HoursRepository hoursRepository, ProjectRepository projectRep
...
-        result.addAll(project.getTeam());
-        result.retainAll(filter.getEmployeesInViewOfDepartments());
-        return result;
-    }
+	protected SortedSet<Employee> getSortedTeamMembers(Project project, List<Hours> hoursList) {
+		SortedSet<Employee> result = hoursList.stream()
+				.map(Hours::getEmployee)
+				.collect(Collectors.toCollection(() -> new TreeSet<>(Employee.NAME_COMPARATOR)));
+		result.addAll(project.getTeam());
+		List<Employee> employees = filter.getEmployeesInViewOfDepartments();
```
Comment (eshestakov) [inline arttime/arttime-webapp/src/main/java/com/artezio/arttime/web/spread_sheet/SpreadSheet.java]: Possible NPE, coming from getEmployeesInViewOfDepartments

File: arttime/arttime-webapp/src/main/java/com/artezio/arttime/web/spread_sheet/SpreadSheet.java [hunk 4]
```diff
@@ -50,13 +59,27 @@ public SpreadSheet(HoursRepository hoursRepository, ProjectRepository projectRep
...
+		if (employees.isEmpty()) {
+			employees = proceedIfEmployeesAreEmpty();
+		}
+		result.retainAll(employees);
+		return result;
+	}
+
+	protected List<Employee> getSortedEmployees() {
+		List<Employee> employees = filter.getEmployeesInViewOfDepartments();
+		if (employees.isEmpty()) {
```
Comment (eshestakov) [inline arttime/arttime-webapp/src/main/java/com/artezio/arttime/web/spread_sheet/SpreadSheet.java]: -//- same

File: arttime/arttime-webapp/src/main/java/com/artezio/arttime/web/spread_sheet/SpreadSheet.java [hunk 5]
```diff
@@ -65,32 +88,35 @@ public SpreadSheet(HoursRepository hoursRepository, ProjectRepository projectRep
...
-		List<Employee> employees = filter.getEmployeesInViewOfDepartments();		
-		employees.sort(Employee.NAME_COMPARATOR);
-		return employees;
+	protected SortedSet<HourType> getSortedHourTypes(Project project, List<Hours> hoursList) {
+		SortedSet<HourType> result = hoursList.stream()
+				.map(Hours::getType)
+				.collect(Collectors.toCollection(() -> new TreeSet<>(HourType.ACTUALTIME_TYPE_COMPARATOR)));
+		result.addAll(project.getAccountableHours());
+		List<HourType> hourTypes = filter.getHourTypes();
+		if (hourTypes.isEmpty()) {
```
Comment (eshestakov) [inline arttime/arttime-webapp/src/main/java/com/artezio/arttime/web/spread_sheet/SpreadSheet.java]: -//- same

File: arttime/arttime-webapp/src/main/java/com/artezio/arttime/filter/Filter.java (hunk 1 above)
Comment (andgeth) [reply arttime/arttime-webapp/src/main/java/com/artezio/arttime/filter/Filter.java]: No, these fields are checked in setters

File: arttime/arttime-webapp/src/main/java/com/artezio/arttime/web/spread_sheet/ProjectEffortsSpreadSheet.java (hunk 2 above)
Comment (andgeth) [reply arttime/arttime-webapp/src/main/java/com/artezio/arttime/web/spread_sheet/ProjectEffortsSpreadSheet.java]: No, answer above

File: arttime/arttime-webapp/src/main/java/com/artezio/arttime/web/spread_sheet/SpreadSheet.java (hunk 3 above)
Comment (andgeth) [reply arttime/arttime-webapp/src/main/java/com/artezio/arttime/web/spread_sheet/SpreadSheet.java]: Same answer

File: arttime/arttime-webapp/src/main/java/com/artezio/arttime/web/spread_sheet/SpreadSheet.java (hunk 4 above)
Comment (andgeth) [reply arttime/arttime-webapp/src/main/java/com/artezio/arttime/web/spread_sheet/SpreadSheet.java]: Same answer

File: arttime/arttime-webapp/src/main/java/com/artezio/arttime/web/spread_sheet/SpreadSheet.java (hunk 5 above)
Comment (andgeth) [reply arttime/arttime-webapp/src/main/java/com/artezio/arttime/web/spread_sheet/SpreadSheet.java]: Same answer

**Pull request merged**